### DISCUSSION TITLE
How do I run Dagster jobs consisting of multiple steps, each with its own set of dependencies and image?

### DISCUSSION THREAD

**Discussion by gibsondan** [dagster-io/dagster#14387] [Q&A] [answered]
Each step of my job needs to be packaged in its own image due to dependencies conflicts and each step needs to be executed either in a separate pod or job - how do I do this?

**Comment by gibsondan** [comment]
There are two ways to do this in Dagster, depending on how your steps are written.

**Option A:**
If you want to use Dagster to orchestrate multiple images that aren't written in Dagster (or even necessarily in Python) then we offer some built-in operators that run a container for you in their op bodies. You can use [`k8s_job_op` ] (https://docs.dagster.io/_apidocs/libraries/dagster-k8s#dagster_k8s.k8s_job_op) to run an arbitrary image in an op in a Kubernetes cluster, or `docker_container_op`(https://docs.dagster.io/_apidocs/libraries/dagster-docker#dagster_docker.docker_container_op) to run an arbitrary image if you have access to a Docker client.

**Option B:**
If you have multiple Dagster ops written in Python and want to run each of them in its own pod, you can vary the [executor](https://docs.dagster.io/deployment/executors#executors) that's used for your job. For example you can use the `k8s_job_executor` to run each op in its own Kubernetes job.

If you also want to vary the image and/or python environment used by each op, then things get trickier. It's possible to vary the image used for each op while using the `k8s_job_executor`, but there are some potential issues that stem from the fact that Dagster needs to be able to load your job definition, both in the run worker that it launches for your run, and the step worker that it launches to run each step:

- If you are using separate python environments that are incompatible with each other, you should delay those incompatible imports until the body of the op that uses them, so that they don't cause an ImportError in the images for your other ops or for the run.

- If you make any changes that would affect your job definition (including inputs or outputs to a particular op) you should update every image being used by your job to reflect that change - otherwise you may get errors or unexpected behavior when running your job.

With those caveats in place, you can alter the image to be used by each op by adding the `dagster-k8s/config` tag to particular ops (or `op_tags` on particular assets) to customize the image used for a particular op when using the `k8s_job_executor`, using the customization tags described [here](https://docs.dagster.io/deployment/guides/kubernetes/customizing-your-deployment#per-job-or-per-op-kubernetes-configuration). For example:

```
from dagster import job, op
from dagster_k8s import k8s_job_executor


@op(
    tags={
        "dagster-k8s/config": {
            "container_config": {"image": "my-custom-image-1:abcde"},
        }
    }
)
def my_custom_image_op(context):
    # Import any dependencies here that don't exist in other images
    ...


@job(executor_def=k8s_job_executor)
def my_custom_image_job():
    my_custom_image_op()
```

We generally recommend only taking this approach if you're sure you need it, since it's more complex and the risk of surprising errors is higher. If you only have a single op that needs to run in a different Python environment or image, Option A above using `k8s_job_op` may be more straightforward,

**Reply by pedrovgp** [reply]
@gibsondan , is there a way to provide these tags dinamically? Our use case is the following:
We are implementing an auto retraining workflow. One of the first assets retrieves a production image. The next asset needs to be run on that image. I would like to use the k8_job_executor, but the image for its configuration will only be available after the first asset runs.

**Reply by dmsfabiano** [reply]
@gibsondan for Option 2, say that we have a specialized ML model image defined somewhere else, and using dagster to run those models, but those images are created else where and do not know about dagster. 

Is there a way to setup the op in the same manner you mentioned in option 2 and somehow now having to define the dagster code in those external images? (I.e pass the op definition to the executor running in the specific image?)

**Reply by gibsondan** [reply]
I'd suggest taking a look at Dagster Pipes for this use case: https://docs.dagster.io/guides/dagster-pipes/kubernetes

**Reply by erenz14** [reply]
@gibsondan I see in all of the pipes guide examples, they are using assets and sending structured metadata from the external code back to the asset via _report_asset_materialization_.

If we wanted to send a json object from the external code, back to an **op**, is there any way to do this?

**Reply by alangenfeld** [reply]
Not currently, would you be up to file a feature request with some more context ? 

edit: technically you could use `report_asset_materialization` to move metadata and process it yourself in some custom way but would be a bit awkward

**Reply by erenz14** [reply]
To respond to your edit, I've actually tried calling _report_asset_materialization_ when the pipes session was opened by an op. But since the context object being used is of type OpExecutionContext, it doesn't have an asset key and so is throwing this error: 
 
Calling `report_asset_materialization` without passing an asset key is undefined. Current step does not target a specific asset.
 
Hopefully I'm just missing something on my end though because this would be a good enough short term solution.
 
And yes, will open a feature request right now. Thanks for the quick response!

**Reply by alangenfeld** [reply]
Ah ya that op issue has a fix coming in the next release https://github.com/dagster-io/dagster/pull/18077

**Reply by alangenfeld** [reply]
> Calling report_asset_materialization without passing an asset key is undefined. Current step does not target a specific asset.

I think if you just fill out a dummy value for `asset_key` that may overcome that hurdle.

**Reply by erenz14** [reply]
Oh I actually have seen that error too, so happy to see its getting fixed. And will try passing a dummy value for asset key, to get around the error I mentioned.
 
Also, I created a feature request here: https://github.com/dagster-io/dagster/issues/18201

**Comment by gibsondan** [comment, marked as answer]
Dagster Pipes is a new feature that can be used to accomplish this: https://docs.dagster.io/guides/dagster-pipes
